## Section 1 - Imports

In [2]:
import pandas as pd # For Data Manipulation of Numerical Yables and Time Series 
import numpy as np # For Mathematical operations on arrays
import sklearn
import warnings
import matplotlib as plt 
import seaborn as sns
import datetime
import glob

## Section 2 - Data loading

In [3]:
patient = pd.read_csv('data/patient.csv', encoding='utf-8', engine='python')
clinical_codes = pd.read_csv('data/clinical_codes.csv', encoding='utf-8', engine='python')

## Section 3a- Visualise Top of Patient table

In [4]:
patient.drop(columns=["Unnamed: 0"], inplace=True, errors='ignore')
patient.head()


age  postcode patient_surname patient_givenname        date_of_birth  \
0    0   HX2 8WG            Bibi           Maureen  2023-02-11 00:00:00   
1    0  LS10 8MH         O'Brien             Karen  2022-11-08 00:00:00   
2    0  WF17 2XU          Turner              Mark  2023-01-28 00:00:00   
3    0  WF13 3RJ             Law         Frederick  2022-12-05 00:00:00   
4    0  LS99 9ZZ          Connor           Emerald  2022-11-17 00:00:00   

  date_of_death                     registration_guid  patient_id  \
0           NaN  9A70C9CD-D0F7-43F3-A05C-34C8F192E00C       10457   
1           NaN  57E3019C-25E6-4591-AA6D-3139AF37DC3F       11488   
2           NaN  A1700F02-134E-42FF-B537-1B7DA00F8C25        9844   
3           NaN  CEDE39B9-9261-4B21-9E07-92D4308E87D5        9972   
4           NaN  9532DEBE-42CC-40E4-820A-55DF603E0D23       18955   

          gender  
0         Female  
1         Female  
2           Male  
3           Male  
4  Indeterminate

## Section 3b - Visualise Bottom of Patient table

In [5]:
patient.tail()

age  postcode patient_surname patient_givenname        date_of_birth  \
4538    0  LS99 9ZZ           Lyons             Ahmir  2023-01-08 00:00:00   
4539    0  LS99 9ZZ      Williamson             Cyrus  2023-02-09 00:00:00   
4540    0  LS99 9ZZ           Healy            Kamari  2022-12-10 00:00:00   
4541    0  LS99 9ZZ          Bolton         Zachariah  2022-11-23 00:00:00   
4542    0  LS99 9ZZ         Johnson          Adrianna  2022-12-05 00:00:00   

     date_of_death                     registration_guid  patient_id  \
4538           NaN  61CE65F5-271B-4151-8700-70DE00CC0A9B       29589   
4539           NaN  4C16CFF4-520C-4B1C-8B03-9B214BDFAC76       19592   
4540           NaN  21107E5F-CE16-4669-9B6E-43671212E9D0       27505   
4541           NaN  9F8F0E53-7252-4998-A035-6809941C8A83       30629   
4542           NaN  F2E5C6F8-52B3-477B-83B2-955A9BB227A5       26947   

             gender  
4538         Female  
4539         Female  
4540  Indeterminate  
4541           Male  
4542         Female

## Section 3 - Descriptive statistics on Patient table

From just looking at the data, it is hard to get a feeling of what all the columns and rows mean. To gain an understanding of the dataset, let's first look at what columns and datatypes we have

In [6]:
patient.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4543 entries, 0 to 4542
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   age                4543 non-null   int64 
 1   postcode           4438 non-null   object
 2   patient_surname    4543 non-null   object
 3   patient_givenname  4543 non-null   object
 4   date_of_birth      4543 non-null   object
 5   date_of_death      13 non-null     object
 6   registration_guid  4543 non-null   object
 7   patient_id         4543 non-null   int64 
 8   gender             4543 non-null   object
dtypes: int64(2), object(7)
memory usage: 319.6+ KB


The `.info()` method creates a data description of the dataframe. It tells us:

- How many rows (4543)
- How many columns (9)
- Column names
- How many values are non-null for each column
- The types of data contained within each column
- The size of the dataset loaded into memory (~319.6KB)

Looking at the output of the `.info()` method, we can intepret each column as follows:

We can see that there are 3 different datatypes within this dataset:

- object = this column contains categorical values
- int64 = this column contains integer values (whole numbers)

In [6]:
patient.describe()

age    patient_id
count  4543.000000   4543.000000
mean      0.331939  16704.930222
std       4.352718   6747.116037
min       0.000000   9596.000000
25%       0.000000  11116.500000
50%       0.000000  12258.000000
75%       0.000000  22793.500000
max      88.000000  30629.000000

The `.describe()` method computes some descriptive statistics of the numerical columns, including:

- count = count of how many unique values exist
- mean = mean average value of this column
- std = standard deviation
- min = minimum value
- 25% = lower quartile value
- 50% = median value
- 75% = upper quartile value
- max = maximum value

We can also see that there are a lot of 0's or nulls effecting the data so we'll fix that next

In [5]:
patientFix = patient[patient != 0]
patientFix.describe().applymap('{:,.2f}'.format) # Convert to 2 decimal places

age patient_id
count  42.00   4,543.00
mean   35.90  16,704.93
std    28.12   6,747.12
min     1.00   9,596.00
25%    10.25  11,116.50
50%    34.00  12,258.00
75%    58.50  22,793.50
max    88.00  30,629.00

## Section 4 - Clinical Codes Table


In [7]:
pd.set_option('display.float_format', lambda x: '%.2f' % x) # Suppress Scientific Notation from Pandas Aggregation Results
clinical_codes.head()

refset_simple_id  parent_code_id            code_id  \
0 999012891000230144.00             NaN   9315231000006113   
1 999012891000230144.00             NaN          301451010   
2 999011571000230144.00             NaN   7627141000006111   
3 999012891000230144.00             NaN  13844351000006118   
4 999011571000230144.00             NaN          506053014   

     snomed_concept_id                                          emis_term  
0 10675991000119100.00  Acute severe exacerbation of mild persistent a...  
1         195949008.00                          Chronic wheezy bronchitis  
2         708030004.00  Pulmonary emphysema co-occurrent with fibrosis...  
3         703954005.00  Asthma caused by house dust mite, Dermatophago...  
4          84409004.00                        Purulent chronic bronchitis

In [9]:
clinical_codes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 586 entries, 0 to 585
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   refset_simple_id   434 non-null    float64
 1   parent_code_id     152 non-null    float64
 2   code_id            586 non-null    int64  
 3   snomed_concept_id  439 non-null    float64
 4   emis_term          586 non-null    object 
dtypes: float64(3), int64(1), object(1)
memory usage: 23.0+ KB


## Section 5 - Joining Medication table
In our data folder we have a medication.csv and 14 more CSV files from medication_2.csv through to medication_15.csv

In [8]:
# read the initial dataframe with 63 columns
medication = pd.read_csv('data/medication/medication.csv', encoding='utf-8', engine='python')

In [10]:
medication.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15096 entries, 0 to 15095
Data columns (total 63 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   nhs_prescribing_agency                    15096 non-null  object 
 1   emis_drug_guid                            0 non-null      float64
 2   authorisedissues_authorised_date          15096 non-null  object 
 3   authorisedissues_first_issue_date         13858 non-null  object 
 4   cancellation_reason                       52 non-null     object 
 5   emis_code_id                              15096 non-null  int64  
 6   confidential_flag                         15096 non-null  bool   
 7   consultation_source_emis_code_id          0 non-null      float64
 8   consultation_source_emis_original_term    0 non-null      float64
 9   dose                                      15096 non-null  object 
 10  emis_medication_status            

In [12]:
medication.tail()

nhs_prescribing_agency  emis_drug_guid  \
15091  prescribed-at-gp-practice             NaN   
15092  prescribed-at-gp-practice             NaN   
15093  prescribed-at-gp-practice             NaN   
15094  prescribed-at-gp-practice             NaN   
15095  prescribed-at-gp-practice             NaN   

      authorisedissues_authorised_date authorisedissues_first_issue_date  \
15091              2001-08-31 00:00:00               2001-08-31 00:00:00   
15092              2000-03-24 00:00:00               2000-03-24 00:00:00   
15093              2001-05-08 00:00:00               2001-05-08 00:00:00   
15094              1999-06-18 00:00:00               2000-03-24 00:00:00   
15095              2001-08-10 00:00:00               2001-08-10 00:00:00   

      cancellation_reason      emis_code_id  confidential_flag  \
15091                 NaN   455941000033116              False   
15092                 NaN   812441000033117              False   
15093                 NaN   770341000033117              False   
15094                 NaN  2643541000033110              False   
15095                 NaN   747441000033113              False   

       consultation_source_emis_code_id  \
15091                               NaN   
15092                               NaN   
15093                               NaN   
15094                               NaN   
15095                               NaN   

       consultation_source_emis_original_term            dose  ...  \
15091                                     NaN   GARGLE AD QDS  ...   
15092                                     NaN      BD TOPICAL  ...   
15093                                     NaN           0.5ML  ...   
15094                                     NaN  TAKE ONE DAILY  ...   
15095                                     NaN     ONE TDS PRN  ...   

                                      emis_original_term  sensitive_flag  \
15091  Difflam Oral Rinse 0.15% solution (Viatris UK ...           False   
15092                  Lasonil 0.8% ointment (Bayer Plc)           False   
15093  Influenza vaccine (split virion, inactivated) ...           False   
15094          Tolterodine 4mg modified-release capsules           False   
15095                            Ibuprofen 400mg tablets           False   

       sensitive_patient_flag    snomed_concept_id snomed_description_id  \
15091                   False   301311000001108.00   2113081000001111.00   
15092                   False  3157411000001100.00  17279401000001116.00   
15093                   False 39716811000001104.00   1815521000001111.00   
15094                   False 14608411000001108.00  50738101000001112.00   
15095                   False         329653008.00  56945501000001120.00   

      fhir_medication_status  cancellation_date nhs_prescription_type  \
15091              completed                NaN                 acute   
15092              completed                NaN                 acute   
15093              completed                NaN                 acute   
15094              completed                NaN                repeat   
15095              completed                NaN                 acute   

              uom uom_dmd  
15091          ml     NaN  
15092     gram(s)     NaN  
15093         NaN     NaN  
15094  capsule(s)     NaN  
15095   tablet(s)     NaN  

[5 rows x 63 columns]

In [9]:
# Lock the columns to 63
medication = medication.iloc[:, :63]
columns = list(medication.columns)
# Define a function to read and append csv files with no headers
def append_csv(file_path):
    df = pd.read_csv(file_path, header=None, engine='python')
    df = df.replace(np.nan, 'placeholder') # Replace all empty values with placeholder
    df = df.iloc[:, :63] # Lock the columns to 63
    df.columns = medication.columns # Set the same column names as the first dataframe
    return df

columns = list(medication.columns)

# Append the remaining csv files
for i in range(2, 16):
    file_path = 'data/medication/medication_{}.csv'.format(i)
    temp = pd.read_csv(file_path, header=None, names=columns)
    medication = medication.append(temp, ignore_index=True)

# Reset the index of the final dataframe
medication = medication.reset_index(drop=True)


In [12]:
medication.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214016 entries, 0 to 214015
Data columns (total 63 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   nhs_prescribing_agency                    214016 non-null  object 
 1   emis_drug_guid                            155340 non-null  object 
 2   authorisedissues_authorised_date          58676 non-null   object 
 3   authorisedissues_first_issue_date         53614 non-null   object 
 4   cancellation_reason                       218 non-null     object 
 5   emis_code_id                              214012 non-null  float64
 6   confidential_flag                         214016 non-null  bool   
 7   consultation_source_emis_code_id          24 non-null      float64
 8   consultation_source_emis_original_term    24 non-null      object 
 9   dose                                      214016 non-null  object 
 10  emis_medication_stat

In [15]:
medication.tail()

nhs_prescribing_agency                        emis_drug_guid  \
214011  prescribed-at-gp-practice  9E0BF639-2120-4BF4-94E7-52D99C0D998A   
214012  prescribed-at-gp-practice  BA5D2A63-0FD6-41C5-BF88-73AD7FC3D40E   
214013  prescribed-at-gp-practice  31061FF7-F81C-4405-B4EF-A5300BEB3897   
214014  prescribed-at-gp-practice  A0A530FA-BA8D-470E-80CF-73230552EF75   
214015  prescribed-at-gp-practice  702601AC-2969-48C8-AAF7-6AF7EDDC3A67   

       authorisedissues_authorised_date authorisedissues_first_issue_date  \
214011                              NaN                               NaN   
214012                              NaN                               NaN   
214013                              NaN                               NaN   
214014                              NaN                               NaN   
214015                              NaN                               NaN   

       cancellation_reason        emis_code_id  confidential_flag  \
214011                 NaN 1454241000033114.00              False   
214012                 NaN 3086141000033116.00              False   
214013                 NaN 1275341000033119.00              False   
214014                 NaN 3038041000033111.00              False   
214015                 NaN  577841000033114.00              False   

        consultation_source_emis_code_id  \
214011                               NaN   
214012                               NaN   
214013                               NaN   
214014                               NaN   
214015                               NaN   

       consultation_source_emis_original_term  \
214011                                    NaN   
214012                                    NaN   
214013                                    NaN   
214014                                    NaN   
214015                                    NaN   

                                     dose  ...  \
214011   ONE TO BE TAKEN FOUR TIMES A DAY  ...   
214012  ONE TO BE TAKEN THREE TIMES A DAY  ...   
214013                          TWO NIGHT  ...   
214014                                 OD  ...   
214015                            2 DIALY  ...   

                               emis_original_term  sensitive_flag  \
214011    Tramadol Hydrochloride  Capsules  50 mg           False   
214012              Chlorphenamine  Tablets  4 mg           False   
214013                     Senna  Tablets  7.5 mg           False   
214014                  Amlodipine  Tablets  5 mg           False   
214015  Fluoxetine Hydrochloride  Capsules  20 mg           False   

        sensitive_patient_flag    snomed_concept_id snomed_description_id  \
214011                   False         322623000.00  46669601000001112.00   
214012                   False 39708211000001104.00   1815991000001119.00   
214013                   False         317548002.00  56949201000001112.00   
214014                   False 39732011000001104.00   1815721000001115.00   
214015                   False         321949006.00  56914201000001120.00   

       fhir_medication_status cancellation_date nhs_prescription_type  \
214011              completed               NaN                 acute   
214012              completed               NaN                repeat   
214013              completed               NaN                repeat   
214014              completed               NaN                repeat   
214015              completed               NaN                repeat   

               uom uom_dmd  
214011  capsule(s)     NaN  
214012   tablet(s)     NaN  
214013      tablet     NaN  
214014      tablet     NaN  
214015  capsule(s)     NaN  

[5 rows x 63 columns]

## Section 5 - Joining observation table
In our data folder we have a observation.csv and 11 more CSV files from observation_2.csv through to observation_12.csv

In [10]:
# read the initial dataframe 
observation = pd.read_csv('data/observation/observation.csv', encoding='utf-8', engine='python')

In [14]:
observation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23186 entries, 0 to 23185
Data columns (total 50 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   abnormal                                  23186 non-null  bool   
 1   emis_code_id                              23186 non-null  int64  
 2   comparator                                1 non-null      object 
 3   confidential_flag                         23186 non-null  bool   
 4   confidential_patient_flag                 23186 non-null  bool   
 5   consultation_source_emis_code_id          10184 non-null  float64
 6   consultation_source_emis_original_term    10184 non-null  object 
 7   document_guid                             1253 non-null   object 
 8   dummy_patient_flag                        23186 non-null  bool   
 9   effective_date                            23166 non-null  object 
 10  effective_date_precision          

In [18]:
observation.tail()

abnormal  emis_code_id comparator  confidential_flag  \
23181     False    1208954011        NaN              False   
23182     False     254063019        NaN              False   
23183     False     264567012        NaN              False   
23184      True    1495289019        NaN              False   
23185     False     259579011        NaN              False   

       confidential_patient_flag  consultation_source_emis_code_id  \
23181                      False               1672871000006114.00   
23182                      False               1672871000006114.00   
23183                      False               1672871000006114.00   
23184                      False               1672871000006114.00   
23185                      False                               NaN   

      consultation_source_emis_original_term document_guid  \
23181                           Main Surgery           NaN   
23182                           Main Surgery           NaN   
23183                           Main Surgery           NaN   
23184               General Practice Surgery           NaN   
23185                                    NaN           NaN   

       dummy_patient_flag       effective_date  ... sensitive_patient_flag  \
23181               False  2001-03-23 00:00:00  ...                  False   
23182               False  2003-12-03 00:00:00  ...                  False   
23183               False  2000-10-08 00:00:00  ...                  False   
23184               False  2004-02-29 00:00:00  ...                  False   
23185               False  2005-04-06 00:00:00  ...                  False   

      snomed_concept_id  snomed_description_id  \
23181         370202007             1208954011   
23182         163020007              254063019   
23183         170656008              264567012   
23184   997531000000108       2572661000000117   
23185  1010601000000105       2564971000000119   

               emis_parent_observation_guid  \
23181                                   NaN   
23182                                   NaN   
23183                                   NaN   
23184                                   NaN   
23185  265A7273-2863-4AAF-BDF7-881DD528BAB2   

                exa_parent_observation_guid     uom uom_ucum user_selected  \
23181                                   NaN     NaN      NaN           NaN   
23182                                   NaN     NaN      NaN           NaN   
23183                                   NaN     NaN      NaN           NaN   
23184                                   NaN     NaN      NaN           NaN   
23185  2455013f-7c79-5b3d-90f9-03fec5de7d9c  mmol/L   mmol/L           NaN   

      numericvalue  value_pq_2  
23181          NaN         NaN  
23182          NaN         NaN  
23183          NaN         NaN  
23184          NaN         NaN  
23185         1.10         NaN  

[5 rows x 50 columns]

In [11]:
# Lock the columns to 50
observation = observation.iloc[:, :50]
columns = list(observation.columns)
# Define a function to read and append csv files with no headers
def append_csv(file_path):
    df = pd.read_csv(file_path, header=None, engine='python', low_memory=False)
    df = df.replace(np.nan, 'placeholder') # Replace all empty values with placeholder
    df = df.iloc[:, :50] # Lock the columns to 50
    df.columns = observation.columns # Set the same column names as the first dataframe
    return df

columns = list(observation.columns)

# Append the remaining csv files
for i in range(2, 13):
    file_path = 'data/observation/observation_{}.csv'.format(i)
    temp = pd.read_csv(file_path, header=None, names=columns)
    observation = observation.append(temp, ignore_index=True)

# Reset the index of the final dataframe
observation = observation.reset_index(drop=True)

c:\Anaconda3\envs\emisenv\lib\site-packages\IPython\core\interactiveshell.py:3552: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [17]:
observation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 528350 entries, 0 to 528349
Data columns (total 50 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   abnormal                                  528350 non-null  bool   
 1   emis_code_id                              528350 non-null  int64  
 2   comparator                                31 non-null      object 
 3   confidential_flag                         528350 non-null  bool   
 4   confidential_patient_flag                 528350 non-null  bool   
 5   consultation_source_emis_code_id          215880 non-null  float64
 6   consultation_source_emis_original_term    215880 non-null  object 
 7   document_guid                             27289 non-null   object 
 8   dummy_patient_flag                        528350 non-null  bool   
 9   effective_date                            527760 non-null  object 
 10  effective_date_preci

In [22]:
observation.tail()

abnormal     emis_code_id comparator  confidential_flag  \
528345     False  150381000006115        NaN              False   
528346     False        291352018        NaN              False   
528347     False        291352018        NaN              False   
528348     False        285168012        NaN              False   
528349      True  238411000006118        NaN              False   

        confidential_patient_flag  consultation_source_emis_code_id  \
528345                      False                               NaN   
528346                      False               1672871000006114.00   
528347                      False               1672871000006114.00   
528348                      False                 87821000000119.00   
528349                      False                               NaN   

       consultation_source_emis_original_term  \
528345                                    NaN   
528346                           Main Surgery   
528347                           Main Surgery   
528348                     Externally Entered   
528349                                    NaN   

                               document_guid  dummy_patient_flag  \
528345                                   NaN               False   
528346                                   NaN               False   
528347                                   NaN               False   
528348  FBD86579-75FA-46D3-8F9F-A3F17ECA9769               False   
528349                                   NaN               False   

             effective_date  ... sensitive_patient_flag snomed_concept_id  \
528345  2005-10-19 00:00:00  ...                  False  1000621000000104   
528346  2000-11-06 00:00:00  ...                  False         105542008   
528347  2003-10-06 00:00:00  ...                  False         105542008   
528348  2004-03-13 00:00:00  ...                  False         185188008   
528349  2007-12-20 00:00:00  ...                  False  1015481000000107   

        snomed_description_id          emis_parent_observation_guid  \
528345       2560271000000118  0EB63C41-9C19-4C9B-AFA9-8C647FBB4F62   
528346              291352018                                   NaN   
528347              291352018                                   NaN   
528348              285168012                                   NaN   
528349       2552661000000115  79B94E73-151D-446E-92DF-E52D6D392D8E   

                 exa_parent_observation_guid  uom uom_ucum user_selected  \
528345  ed039901-7f10-59a7-b85c-29e56163dcd2  u/L      NaN           NaN   
528346                                   NaN  NaN      NaN           NaN   
528347                                   NaN  NaN      NaN           NaN   
528348                                   NaN  NaN      NaN           NaN   
528349  bbc01ea6-cf08-5f57-adaf-519bd9192c39    %        %           NaN   

       numericvalue  value_pq_2  
528345        51.00         NaN  
528346          NaN         NaN  
528347          NaN         NaN  
528348          NaN         NaN  
528349        12.00         NaN  

[5 rows x 50 columns]

## Section 6 - Finding Patients who meet condition criteria


In [12]:
# Filter patients within LS postcode
ls_postcode_patients = patient[patient['postcode'].fillna('').str.startswith('LS')]
display(ls_postcode_patients)

age  postcode patient_surname patient_givenname        date_of_birth  \
1       0  LS10 8MH         O'Brien             Karen  2022-11-08 00:00:00   
4       0  LS99 9ZZ          Connor           Emerald  2022-11-17 00:00:00   
6       0   LS2 5YB            Daly            Philip  2022-12-19 00:00:00   
7       0  LS14 0HR            Hall            Angela  2022-12-12 00:00:00   
8       0  LS99 9ZZ          Buxton           Isabela  2022-11-10 00:00:00   
...   ...       ...             ...               ...                  ...   
4538    0  LS99 9ZZ           Lyons             Ahmir  2023-01-08 00:00:00   
4539    0  LS99 9ZZ      Williamson             Cyrus  2023-02-09 00:00:00   
4540    0  LS99 9ZZ           Healy            Kamari  2022-12-10 00:00:00   
4541    0  LS99 9ZZ          Bolton         Zachariah  2022-11-23 00:00:00   
4542    0  LS99 9ZZ         Johnson          Adrianna  2022-12-05 00:00:00   

     date_of_death                     registration_guid  patient_id  \
1              NaN  57E3019C-25E6-4591-AA6D-3139AF37DC3F       11488   
4              NaN  9532DEBE-42CC-40E4-820A-55DF603E0D23       18955   
6              NaN  DB724446-0CBB-4BE4-8F8A-AD122804F424       12194   
7              NaN  D53C1CDD-20C4-436F-82D4-D83A60AFDB62        9680   
8              NaN  CD72090A-7FC7-4B20-A9FD-60D52EFD5875       28650   
...            ...                                   ...         ...   
4538           NaN  61CE65F5-271B-4151-8700-70DE00CC0A9B       29589   
4539           NaN  4C16CFF4-520C-4B1C-8B03-9B214BDFAC76       19592   
4540           NaN  21107E5F-CE16-4669-9B6E-43671212E9D0       27505   
4541           NaN  9F8F0E53-7252-4998-A035-6809941C8A83       30629   
4542           NaN  F2E5C6F8-52B3-477B-83B2-955A9BB227A5       26947   

             gender  
1            Female  
4     Indeterminate  
6              Male  
7            Female  
8              Male  
...             ...  
4538         Female  
4539         Female  
4540  Indeterminate  
4541           Male  
4542         Female  

[2640 rows x 9 columns]

In [13]:
# Join the dataframes
df1 = pd.merge(patient, medication, on='registration_guid')

In [14]:
df1 = pd.merge(df1, observation, on='registration_guid')

In [20]:
df1 = pd.merge(df1, clinical_codes, on='code_id')

KeyError: 'code_id'

In [19]:
print(df1.columns)
print(clinical_codes.columns)

Index(['age', 'postcode', 'patient_surname', 'patient_givenname',
       'date_of_birth', 'date_of_death', 'registration_guid', 'patient_id',
       'gender', 'nhs_prescribing_agency',
       ...
       'sensitive_patient_flag_y', 'snomed_concept_id_y',
       'snomed_description_id_y', 'emis_parent_observation_guid',
       'exa_parent_observation_guid', 'uom_y', 'uom_ucum', 'user_selected',
       'numericvalue', 'value_pq_2'],
      dtype='object', length=120)
Index(['refset_simple_id', 'parent_code_id', 'code_id', 'snomed_concept_id',
       'emis_term'],
      dtype='object')


In [21]:
# Filter the patients within the LS postcode
ls_postcode_df = df1[df1['postcode'] == 'LS']

MemoryError: Unable to allocate 8.73 GiB for an array with shape (20, 58604527) and data type float64

In [21]:
# Get the relevant clinical codes for smoker refset
smoker_codes = clinical_codes[clinical_codes['refset_simple_id'] == 999004211000230104]['code_id'].tolist()

In [22]:
# Get the relevant clinical codes for medication ingredients
medication_codes = clinical_codes[(clinical_codes['parent_code_id'].isin([591221000033116, 717321000033118, 1215621000033114, 972021000033115, 1223821000033118])) & (clinical_codes['snomed_concept_id'].isin([129490002, 108606009, 702408004, 702801003, 704459002]))]['code_id'].tolist()

In [83]:
# Step 2: Filter clinical codes with parent_code_id or snomed_concept_id specified.
parent_codes = [591221000033116, 717321000033118, 1215621000033114, 972021000033115, 1223821000033118]
snomed_concepts = [129490002, 108606009, 702408004, 702801003, 704459002]
clinical_codes_filtered = clinical_codes[(clinical_codes['refset_simple_id'] == 999012891000230104) & ((clinical_codes['parent_code_id'].isin(parent_codes)) | (clinical_codes['snomed_concept_id'].isin(snomed_concepts))) & (~clinical_codes['refset_simple_id'].isin([999004211000230104, 999011571000230107]))]

In [86]:
# Step 3: Filter observations and medications with dummy_patient_flag and confidential_patient_flag set to false.
obs_med_filtered = pd.concat([observation[(observation['dummy_patient_flag'] == False) & (observation['confidential_patient_flag'] == False) & (observation['confidential_flag'] == False)], medication[(medication['dummy_patient_flag'] == False) & (medication['confidential_flag'] == False)]])

In [48]:
# filter observations and medications with dummy_patient_flag and confidential_patient_flag set to False
observation = observation[(observation["dummy_patient_flag"] == False) & (observation["confidential_patient_flag"] == False)& (observation["confidential_flag"] == False)]
medication = medication[(medication["dummy_patient_flag"] == False) & (medication["confidential_flag"] == False)]

In [89]:
# Step 4: Merge all filtered data frames and get the required columns.
result = pd.merge(ls_patients[['registration_guid', 'patient_id', 'patient_surname', 'patient_givenname', 'postcode', 'date_of_birth', 'gender']], obs_med_filtered[['registration_guid', 'emis_code_id', 'snomed_concept_id']], on='registration_guid')
result = pd.merge(result, clinical_codes_filtered[['refset_simple_id', 'code_id', 'snomed_concept_id']], left_on='emis_code_id', right_on='code_id')
result = result.drop_duplicates(subset=['registration_guid', 'patient_id'])
result = result[['registration_guid', 'patient_id', 'patient_surname', 'patient_givenname', 'postcode', 'date_of_birth', 'gender']]

In [90]:
print(result)

Empty DataFrame
Columns: [registration_guid, patient_id, patient_surname, patient_givenname, postcode, date_of_birth, gender]
Index: []


In [80]:
# Split the patient DataFrame into smaller chunks based on the postcode column
postcode_chunks = patient['postcode'].fillna('').str.slice(0, 2).unique().tolist()
patient_chunks = [patient[patient['postcode'].fillna('').str.startswith(p)] for p in postcode_chunks]

# Process each chunk separately and concatenate the results
ls_result_chunks = []
for chunk in patient_chunks:
    ls_patients = chunk.merge(observation, on="registration_guid", how="inner")
    ls_patients = ls_patients.merge(medication, on="registration_guid", how="inner")
    ls_patients = ls_patients[(ls_patients["refset_simple_id"] == 999012891000230104) & 
                              (~ls_patients["refset_simple_id"].isin([999004211000230104, 999011571000230107])) & 
                              (~ls_patients["snomed_concept_id"].isin([27113001])) & 
                              ((ls_patients["parent_code_id"].isin([591221000033116, 717321000033118, 1215621000033114, 972021000033115, 1223821000033118])) | 
                               (ls_patients["snomed_concept_id"].isin([129490002, 108606009, 702408004, 702801003, 704459002]))) & \
                       (ls_patients['postcode'].fillna('').str.startswith('LS'))]
    ls_result = ls_patients[["registration_guid", "patient_id", "patient_surname", "patient_givenname", "postcode", "age", "gender"]]
    ls_result_chunks.append(ls_result)

ls_result = pd.concat(ls_result_chunks)

KeyError: 'refset_simple_id'

In [33]:
print("Patient columns:", list(patient.columns))
print("Medication columns:", list(medication.columns))
print("Observation columns:", list(observation.columns))
print("Clinical codes columns:", list(clinical_codes.columns))

Patient columns: ['age', 'postcode', 'patient_surname', 'patient_givenname', 'date_of_birth', 'date_of_death', 'registration_guid', 'patient_id', 'gender']
Medication columns: ['nhs_prescribing_agency', 'emis_drug_guid', 'authorisedissues_authorised_date', 'authorisedissues_first_issue_date', 'cancellation_reason', 'emis_code_id', 'confidential_flag', 'consultation_source_emis_code_id', 'consultation_source_emis_original_term', 'dose', 'emis_medication_status', 'dummy_patient_flag', 'duration_in_days', 'duration_uom', 'effective_date', 'effective_date_precision', 'emis_issue_method', 'emis_mostrecent_issue_date', 'emis_prescription_type', 'emis_registration_organisation_guid', 'emis_encounter_guid', 'exa_encounter_guid', 'end_date', 'emis_enteredby_userinrole_guid', 'exa_prescription_guid', 'estimated_nhs_cost', 'exa_drug_guid', 'medication_guid', 'exa_medication_guid', 'fhir_medication_intent', 'exa_mostrecent_issue_date', 'max_nextissue_days', 'min_nextissue_days', 'non_regular_and_c